### PR TITLE
JIT: merge fgMorph into compCompile

### PR DESCRIPTION
**Pull request by AndyAyersMS**
This gives us a single method that controls most of the jit's phase behavior.
Largely a manual inline, though I updated some comments and changed one assert.

Follow up from #1309; also see #2109.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@dotnet/jit-contrib PTAL

File: src/coreclr/src/jit/compiler.cpp [hunk 1]
```diff
@@ -4226,7 +4230,8 @@ void Compiler::compCompile(void** methodCodePtr, ULONG* methodCodeSize, JitFlags
         return;
     }
 
-    assert(!compDonotInline());
+    // By this point we are only jitting the root method (and not an inlinee)
```
Comment (CarolEidt) [inline src/coreclr/src/jit/compiler.cpp]: Minor: I might reword this as "By this point we have completed inlining into the root method, so going forward we won't encounter any inlinees" or something like that.

File: src/coreclr/src/jit/compiler.cpp [hunk 2]
```diff
@@ -4575,18 +4767,18 @@ void Compiler::compCompile(void** methodCodePtr, ULONG* methodCodeSize, JitFlags
...
+    // Assign registers to variables, etc.
 
     ///////////////////////////////////////////////////////////////////////////////
     // Dominator and reachability sets are no longer valid. They haven't been
     // maintained up to here, and shouldn't be used (unless recomputed).
     ///////////////////////////////////////////////////////////////////////////////
     fgDomsComputed = false;
 
-    /* Create LSRA before Lowering, this way Lowering can initialize the TreeNode Map */
+    // Create LSRA before Lowering, this way Lowering can initialize the TreeNode Map
```
Comment (CarolEidt) [inline src/coreclr/src/jit/compiler.cpp]: Nit: since you're here, the TreeNode map is gone; I would reword this as something like  "Create LinearScan before Lowering, so that Lowering can call LinearScan methods for determining whether locals are register candidates and (for xarch) whether a node is a containable memory op.

File: src/coreclr/src/jit/compiler.cpp [hunk 3]
```diff
@@ -4269,16 +4274,213 @@ void Compiler::compCompile(void** methodCodePtr, ULONG* methodCodeSize, JitFlags
 
     EndPhase(PHASE_POST_IMPORT);
 
-    /* Initialize the BlockSet epoch */
-
+    // Initialize the BlockSet epoch
     NewBasicBlockEpoch();
 
-    /* Massage the trees so that we can generate code out of them */
+    // Start of the code that used to make up fgMorph
```
Comment (CarolEidt) [inline src/coreclr/src/jit/compiler.cpp]: I'd avoid mentioning what used to be, and instead encompass the comment below, to say something like: "Start of the code that is broadly called morphing, and includes global morph, as well as other phases that massage the trees so that we can generate code out of them."

**Review by CarolEidt — APPROVED** [review]
LGTM, with minor comment suggestions

File: src/coreclr/src/jit/compiler.cpp (hunk 1 above)
Comment (AndyAyersMS) [reply src/coreclr/src/jit/compiler.cpp]: At this point we haven't started inlining into the root method.

We run a few of the initial phases on inlinees. So inlinees and root methods both call `compCompile`. This comment is trying to say that "at this point in the phase list, all the inlinee phases have been run, and inlinee compiles have exited, so we should only get this far if we are jitting the root method."

If that quoted text seems better, let me know, and I'll update...

Comment (CarolEidt) [reply src/coreclr/src/jit/compiler.cpp]: Yes, that would be great - and thanks for clarifying!

File: src/coreclr/src/jit/compiler.cpp (hunk 3 above)
Comment (AndyAyersMS) [reply src/coreclr/src/jit/compiler.cpp]: Updated

File: src/coreclr/src/jit/compiler.cpp (hunk 2 above)
Comment (AndyAyersMS) [reply src/coreclr/src/jit/compiler.cpp]: Updated

**Comment by AndyAyersMS** [comment]
Failures almost certainly unrelated.

**Review by BruceForstall — APPROVED** [review]
Seems like we could simplify this even further as a further change by sinking checks and #ifdefs into phase / function calls, extracting functions for things like compGcChecks, etc., so the function ends up being almost entirely function calls to phase-like functions and `EndPhase` calls.

**Comment by AndyAyersMS** [comment]
> Seems like we could simplify this even further

That's where I am headed (in gradual steps), see #2109...

File: src/coreclr/src/jit/compiler.cpp [hunk 4]
```diff
@@ -4129,16 +4129,24 @@ void Compiler::compFunctionTraceEnd(void* methodCodePtr, ULONG methodCodeSize, b
...
 }
 
-//*********************************************************************************************
-// #Phases
+//------------------------------------------------------------------------
+// compCompile: run phases needed for compilation
+//
+// Arguments:
+//   methodCodePtr [OUT] - address of generated code
+//   methodCodeSize [OUT] - size of the generated code (hot + cold setions)
```
[the remark this answers is not in the document]
Comment (AndyAyersMS) [reply src/coreclr/src/jit/compiler.cpp]: Thanks; will fix this in subsequent change.
